[DEVELOP] Adding TAD class

gwaybio opened the pull request:
Adding method to facilitate a TAD_Pathways analysis. The method calls all the scripts in order of the tad_pathways pipeline given a file storing SNP info.

This is in preparation of Diana's Longitudinal BMD paper. Not really sure of where this should all live... 

The TAD_Pathways portion of the longitudinal BMD analysis consists of three steps:

1. Running TAD_Pathways on 41 different SNP lists
2. Simulating gene lists using randomly selected TADs matched by gene content similarity
3. Comparing implicated genes in real TAD_Pathways to permuted gene lists (and Diana digs deeper into these specific genes in downstream analyses)

At this point, nearly all of the functionality lives on the `develop` branch in this repo. All of it has been reviewed already (except this PR) (see #21, #22, #25, #26, #30, #31, #33). My original goal was to make this repo a python package so that Diana's analysis could live in an independent repo - but this goal is likely to require several more hours of work. 

An alternative is to add the longitudinal BMD analysis here in a separate branch (can branch off of `develop`) and generate a release on this specific branch. I think this will take the least amount of time.

There is also the element of code review. I certainly don't want to take up any time of anyone in the lab...

Thoughts @cgreene ?

gwaybio committed:
add tad.py

gwaybio committed:
add config python method

gwaybio committed:
add genelist method

gwaybio committed:
run black on added files

cgreene commented:
For all of the reasons that code review is important including the opportunity to catch certain bugs, identify new solutions, and making sure code is reusable across the lab (and hopefully also the field), it seems worth going through review. This isn't an inordinately large amount of code. I bet that if you asked @dongbohu he would be happy to take a look.

I'm fine with the develop branch of this repo. You might also be able to get help from @dongbohu to make this repo into a python package. I'm not sure if he has the capacity but it seems like it'd be worth asking.

dongbohu commented on tad_pathways/config.py:
```diff
@@ -0,0 +1,31 @@
+import yaml
+
+
+def config_yaml(config_file):
```
So I suppose this module will be used in the future by users to read in a yaml config file?

dongbohu commented on tad_pathways/config.py:
```diff
@@ -0,0 +1,31 @@
+    # Load configuration
+    with open(config_file, "r") as stream:
+        config = yaml.load(stream)
+
```
Do you have a template yaml file? I guess some error handling logic will be needed in the future to validate the input yaml file.

gwaybio replied:
This isn't currently used, but will be in the future. I will add error handling in a future PR (see #35)

dongbohu commented on tad_pathways/config.py:
```diff
@@ -0,0 +1,31 @@
+"""
+Gregory Way 2018
```
2019?

dongbohu commented on tad_pathways/genes/genelist.py:
```diff
@@ -0,0 +1,167 @@
+"""
+Gregory Way 2018
```
2019?

dongbohu commented on tad_pathways/genes/genelist.py:
```diff
@@ -0,0 +1,167 @@
+"""
+Gregory Way 2018
+TAD Pathways
+scripts/genes/genelist.py
```
This path doesn't match the actual location.

dongbohu commented on tad_pathways/genes/genelist.py:
```diff
@@ -0,0 +1,167 @@
+
+class TADGeneList:
+    """
+    Class methods to build a genelist for TAD associated genes
```
You may want to delete the word "methods" here, because `class methods` usually refers to `@classmethod`.

dongbohu commented on tad_pathways/genes/genelist.py:
```diff
@@ -0,0 +1,167 @@
+              The SNP DataFrame has the columns: [RSid, database source, chromosome,
+              genomic position, minor allele, and group]
+        gene_index_file - file location of TAD index file
+        remove_hla - boolean if HLA TAD should be removed
```
`remove_hla_tad`?

dongbohu commented on tad_pathways/genes/genelist.py:
```diff
@@ -0,0 +1,167 @@
+
+        return tad.reset_index(drop=True)
+
+    def compile(self, remove_hla_tad=False):
```
I wonder whether `remove_hla_tad` parameter is necessary. According to line 100:
```
if remove_hla_tad or self.remove_hla_tad:
    ...
```
suppose a user creates an instance of this class by:
```
x = TADGeneList(..., remove_hla_tad=True)
```
then call:
```
x.compile(remove_hla_tad=False)
```
the user may intend to NOT remove hla_tad, but because `self.remove_hla_tad` is True, it will still be done. 

(I don't know exactly how this class will be used, so this comment may not make sense at all.)

gwaybio replied:
Ah, great point! I think the best course is to drop the `remove_hla_tad` argument from `compile()`

dongbohu commented on tad_pathways/genes/genelist.py:
```diff
@@ -0,0 +1,167 @@
+                    continue
+
+                if remove_hla_tad or self.remove_hla_tad:
+                    if tad_id[0] == "1198":
```
`1198` is hard coded here. Is it better to create a constant at class or module level?

gwaybio replied:
Yep, I agree. I will make constant at class level

dongbohu commented on tad_pathways/genes/genelist.py:
```diff
@@ -0,0 +1,167 @@
+                    (protein_coding.stop - snp_series["position"]).abs().sort_values()
+                )
+
+                # Closest to the end of a the gene
```
Is this comment supposed to be for line 117?

dongbohu commented on tad_pathways/genes/genelist.py:
```diff
@@ -0,0 +1,167 @@
+
+With the following required flags:
+
+    --snp_data_file     The file name of where SNP data is stored. SNP data is
```
How are these three flags supported? Are you going to add a 
```
if __name__ == "__main__":
```
block to handle them?

gwaybio replied:
This looks to me like deprecated documentation - I will remove!

dongbohu commented on tad_pathways/tad.py:
```diff
@@ -0,0 +1,189 @@
+"""
+Gregory Way 2018
```
2019?

dongbohu commented on tad_pathways/tad.py:
```diff
@@ -0,0 +1,189 @@
+import os
+import subprocess
+
+from tad_pathways.genes.genelist import TADGeneList
```
`TadGeneList` is not being used in this module yet. I suppose it will be in the future?

gwaybio replied:
Good catch - I will remove. It used to be used, but no longer

dongbohu commented on tad_pathways/tad.py:
```diff
@@ -0,0 +1,189 @@
+
+class TadPathways:
+    """
+    Class methods to perform a tad_pathways analysis
```
Same thing about the misleading `class methods`.

dongbohu commented on tad_pathways/tad.py:
```diff
@@ -0,0 +1,189 @@
+        Arguments:
+        snp_list_name - the name of the input snp list
+        base_dir - the root directory of where to save files
+        remove_hla - boolean if HLA TAD should be removed
```
Do you need to explain `snp_list_location` and `all_pathways` arguments too?

gwaybio replied:
thanks for pointing this out - this required a change. I had to update the `pathways` argument in `scripts/construct_evidence.py` and change the way `tad.py` handles `all_pathways`.

I also updated the documentation

dongbohu replied:
:+1:

dongbohu commented on tad_pathways/tad.py:
```diff
@@ -0,0 +1,189 @@
+
+    def build_custom_tad_genelist(self):
+        command_list = [
+            "python",
```
Is `scripts/build_custom_tad_genelist.py` compatible for Python 2 or 3 (or both)? By default, `python` command calls python 2 on many (if not most) operating systems.

gwaybio replied:
python 3 - should I change the call to `"python3"`?

dongbohu replied:
If this package is for Python 3 only, then it is a good idea to change it to `python3`,

gwaybio replied:
done in fbe303d09912f75b79f408e6a9bc0a0419bf134a

dongbohu commented on tad_pathways/tad.py:
```diff
@@ -0,0 +1,189 @@
+        output_pval_file = os.path.join(
+            self.base_dir, "{}_pvals.tsv".format(self.snp_list_name)
+        )
+        if os.path.exists(output_pval_file):
```
If you want, lines 136-139 can be simplified as:
```
return os.path.exists(output_pval_file)
```

gwaybio replied:
:tada:

dongbohu commented on tad_pathways/tad.py:
```diff
@@ -0,0 +1,189 @@
+    def get_evidence(self):
+        # The first command builds the evidence
+        command_list = [
+            "python",
```
Same question on Python 2/3 compatibility.

dongbohu commented on tad_pathways/tad.py:
```diff
@@ -0,0 +1,189 @@
+
+        # The second command summarizes this evidence
+        command_list = [
+            "python",
```
Same question on python 2/3 compatibility.

dongbohu commented on tad_pathways/tad.py:
```diff
@@ -0,0 +1,189 @@
+        webgestalt_exists = self.check_webgestalt()
+        if webgestalt_exists:
+            self.get_evidence()
+            return 1
```
Just curious, the return value of this method will be used later?

gwaybio replied:
Hmm, doesn't look like I use this.. Not sure why its here. Maybe some functionality that I decided not to use. I will remove for now

dongbohu reviewed (commented):
@gwaygenomics: Most of the code looks good. There are a few minor issues for you to address or clarify.

DCousminer commented:
Proposed next steps to complete the enrichment analysis:

1. Output from step 2: divvy into 100 sets of 41 matched TADs
2. Filter by protein-coding genes
3. Submit to DAVID for functional annotation 
4. Count number of times genes appear with specific key words of interest (e.g. "osteoblast"), being careful to count each gene only once
5. Create null distribution based on 100 iterations
6. Compare actual and calculate p-value for enrichment of "real" list vs. simulated

We are working on this-- any thoughts before we proceed are welcome.

gwaybio committed:
respond to PR comments

update documentation, update all_pathways argument, streamline return statements

gwaybio committed:
respond to pr comments

update documentation, remove remove_hla_tad argument from compile()

gwaybio committed:
fix year

gwaybio committed:
add keyerror handling

gwaybio committed:
fix documentation in construct_evidence

gwaybio committed:
change to python3 call

gwaybio commented:
@dongbohu - ready again for re-review!

gwaybio committed:
minor updates to flags

gwaybio committed:
use readr::write_tsv() and fix webgestalt updates

gwaybio committed:
fix for webgestalt updates

dongbohu reviewed (approved):
Looks good!
